File: Chapter8/reverseSLL.js
function Node(value){
    this.val = value;
    this.next = null;
}

function SLL(){
    this.head = null;

    this.add = function(value) {
        var newNode = new Node(value);
        if(!this.head) {this.head = newNode;}
        else {
            var runner = this.head;
            while (runner.next){
                runner = runner.next;
            }
            runner.next = newNode;
        }
        return this;
    }

    this.reverse = function(){
        var prev = null;
        var current = this.head;
        var follow = this.head.next;

        while(follow){
            current.next = prev;
            prev = current;
            current = follow;
            follow = follow.next;
        }
        current.next = prev; // to connect last node to previous node so we don't lose the entire list
        this.head.next = null;
        this.head = current;
        return this;
    }

    this.printAsArray = function(){
        var runner = this.head;
        var arr = [];
        while (runner){
            arr.push(runner.val);
            runner = runner.next;
        }
        console.log(arr);
        return this;
    }
}

var sList = new SLL();
sList.add(1).add(2).add(3).add(4).add(5).printAsArray();
sList.reverse().printAsArray();

File: Chapter8-LINKED_LISTS_PART_II/reverseSLL.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

class SLL:
    def __init__(self):
        self.head = None

    def add(self, val):
        newNode = Node(val)
        if not self.head:
            self.head = newNode
        else:
            runner = self.head
            while runner.next:
                runner = runner.next

            runner.next = newNode
        return self

    def display(self):
        arr = []
        runner = self.head
        while runner:
            arr.append(runner.val)
            runner = runner.next

        print(arr)
        return self

    def reverse(self):
        if not self.head:
            return self
        else:

            prev = self.head
            curr = self.head.next
            next_node = self.head.next.next

            prev.next = None
            curr.next = prev

            while next_node.next:
                prev = curr
                curr = next_node
                next_node = next_node.next
                curr.next = prev

            next_node.next = curr
            self.head = next_node
            return self



s1 = SLL()
s1.add(10).add(9).add(8).add(7).add(6).add(5).add(4).add(3).add(2).add(1).display()
s1.reverse().display()
